src/main/java/org/cryptomator/ui/health/CheckExecutor.java

```diff
@@ -0,0 +1,113 @@
+package org.cryptomator.ui.health;
+
+import org.cryptomator.common.vaults.Vault;
+import org.cryptomator.cryptofs.VaultConfig;
+import org.cryptomator.cryptofs.health.api.DiagnosticResult;
+import org.cryptomator.cryptolib.api.CryptorProvider;
+import org.cryptomator.cryptolib.api.Masterkey;
+
+import javax.inject.Inject;
+import javafx.concurrent.Task;
+import java.nio.file.Path;
+import java.security.SecureRandom;
+import java.util.ArrayDeque;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicReference;
+
+@HealthCheckScoped
+public class CheckExecutor {
+
+	private final Path vaultPath;
+	private final SecureRandom csprng;
+	private final Masterkey masterkey;
+	private final VaultConfig vaultConfig;
+	private final ExecutorService sequentialExecutor;
+	private final Queue<CheckTask> runningTasks;
+
+
+	@Inject
+	public CheckExecutor(@HealthCheckWindow Vault vault, AtomicReference<Masterkey> masterkeyRef, AtomicReference<VaultConfig> vaultConfigRef, SecureRandom csprng) {
+		this.vaultPath = vault.getPath();
+		this.masterkey = masterkeyRef.get();
+		this.vaultConfig = vaultConfigRef.get();
+		this.csprng = csprng;
+		this.sequentialExecutor = Executors.newSingleThreadExecutor();
+		this.runningTasks = new ArrayDeque<>();
+	}
+
+	public synchronized void executeBatch(List<Check> checks) {
+		checks.stream().map(c -> {
+			c.setState(Check.CheckState.SCHEDULED);
+			var task = new CheckTask(c);
+			runningTasks.add(task); // we need to use CheckTask and not Futures to set state to CANCEL
+			return task;
+		}).forEach(sequentialExecutor::submit);
+	}
+
+	public synchronized void cancel() {
+		while (!runningTasks.isEmpty()) {
+			runningTasks.remove().cancel(true);
+		}
```

Comment:
I believe we should iterate backwards.

Otherwise, since we don't shut down the executor, a (worst case) situation may occur where the "cancel thread" C and the "executor thread" E play a turn-based game, where each task is first started by E and immediately cancelled by C.

Reply:
Fixed in b4a530b